File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/reflections/ReflectorExample.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 

Imports System
Imports com.db4o.reflect
Imports com.db4o.reflect.net
Imports com.db4o.reflect.generic
Imports com.db4o.query
Imports com.db4o
Imports System.IO


Namespace com.db4odoc.f1.reflections
    Public Class ReflectorExample
        Public Shared ReadOnly YapFileName As String = "formula1.yap"
        
        Public Sub New()

        End Sub

        Public Sub Main()
            setCars()
            getReflectorInfo()
            getCars()
            getCarInfo()
        End Sub
        ' end Main

        Public Sub SetCars()
            File.Delete(YapFileName)
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim car1 As Car = New Car("BMW")
                db.Set(car1)
                Dim car2 As Car = New Car("Ferrari")
                db.Set(car2)

                Console.WriteLine("Saved:")
                Dim query As Query = db.Query()
                query.Constrain(GetType(Car))
                Dim results As ObjectSet = query.Execute()
                ListResult(results)
            Finally
                db.Close()
            End Try
        End Sub
        ' end SetCars

        Public Sub GetCars()
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim query As Query = db.Query()
                query.Constrain(GetType(Car))
                Dim result As ObjectSet = query.Execute()
                ListResult(result)
                Dim car As Car = CType(result(0), Car)
                Dim reflector As GenericReflector = New GenericReflector(Nothing, db.Ext().Reflector())
                Dim carClass As ReflectClass = reflector.ForObject(car)
                Console.WriteLine("Reflected class " + carClass.GetName())
            Finally
                db.Close()
            End Try
        End Sub
        ' end GetCars

        Public Sub GetCarInfo()
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim result As ObjectSet = db.Get(New Car("BMW"))
                If result.Size() < 1 Then
                    Return
                End If
                Dim car As Car = CType(result(0), Car)
                Dim reflector As GenericReflector = New GenericReflector(Nothing, db.Ext().Reflector())
                Dim carClass As ReflectClass = reflector.ForObject(car)
                Console.WriteLine("Reflected class " + carClass.GetName())
                ' public fields
                Console.WriteLine("FIELDS:")
                Dim fields() As ReflectField = carClass.GetDeclaredFields()
                Dim i As Integer
                For i = 0 To fields.Length - 1 Step i + 1
                    Console.WriteLine(fields(i).GetName())
                Next

                ' constructors
                Console.WriteLine("CONSTRUCTORS:")
                Dim cons() As ReflectConstructor = carClass.GetDeclaredConstructors()
                For i = 0 To cons.Length - 1 Step i + 1
                    Console.WriteLine(cons(i))
                Next

                ' public methods
                Console.WriteLine("METHODS:")
                Dim method As ReflectMethod = carClass.GetMethod("ToString", Nothing)
                Console.WriteLine("ToString method " + CType(method, Object).ToString())
            Finally
                db.Close()
            End Try
        End Sub
        ' end GetCarInfo

        Public Sub GetReflectorInfo()

            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim ref As Reflector
                ref = db.Ext().Reflector()
                Console.WriteLine("Reflector in use: " + CType(ref, Object).ToString())
                Console.WriteLine("Reflector delegate" + CType(db.Ext().Reflector().GetDelegate(), Object).ToString())
                Dim knownClasses As ReflectClass()
                knownClasses = db.Ext().Reflector().KnownClasses()
                Dim count As Integer
                count = knownClasses.Length
                Console.WriteLine("Known classes: " + count.ToString())
                Dim i As Integer
                For i = 0 To knownClasses.Length - 1 Step i + 1
                    Console.WriteLine(knownClasses(i).GetName())
                Next

            Finally
                db.Close()
            End Try
        End Sub
        ' end GetReflectorInfo

        Public Sub TestReflector()
            Dim logger As LoggingReflector = New LoggingReflector()
            Db4o.Configure().ReflectWith(logger)
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim car As Car = New Car("BMW")
                Dim rc As ReflectClass
                rc = db.Ext().Reflector().ForObject(car)
                Console.WriteLine("Reflected class: " + rc.GetName())
            Finally
                db.Close()
            End Try
        End Sub
        ' end TestReflector

        Public Shared Sub ListResult(ByVal result As ObjectSet)
            Console.WriteLine(result.Count)
            For Each item As Object In result
                Console.WriteLine(item)
            Next
        End Sub
        ' end ListResult

    End Class
End Namespace

File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/reflections/ReflectorExample.cs
/* Copyright (C) 2005   db4objects Inc.   http://www.db4o.com */

using System;
using com.db4odoc.f1;
using com.db4o.reflect;
using com.db4o.reflect.net;
using com.db4o.reflect.generic;
using com.db4o.query;
using com.db4o;
using System.IO;

namespace com.db4odoc.f1.reflections
{
	public class ReflectorExample
	{
		public readonly static string YapFileName = "formula1.yap";

		public ReflectorExample()
		{
		}

		public void Main()
		{
			SetCars();
			GetReflectorInfo();
			GetCars();
			GetCarInfo();
		}
		// end Main

		public void SetCars()
		{
			File.Delete(YapFileName);     
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				Car car1 = new Car("BMW");
				db.Set(car1);
				Car car2 = new Car("Ferrari");
				db.Set(car2);
				
				Console.WriteLine("Saved:");
				Query query = db.Query();
				query.Constrain(typeof(Car));
				ObjectSet results = query.Execute();
				ListResult(results);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end SetCars
		
		public  void GetCars()
		{
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				Query query = db.Query();
				query.Constrain(typeof(Car));
				ObjectSet result = query.Execute();
				ListResult(result);
				Car car = (Car)result[0];
				GenericReflector reflector = new GenericReflector(null,db.Ext().Reflector());
				ReflectClass carClass = reflector.ForObject(car);
				Console.WriteLine("Reflected class "+carClass);
				Console.WriteLine("Retrieved with reflector:");
			} 
			finally 
			{
				db.Close();
			}
		}
		// end GetCars
		
		public  void GetCarInfo()
		{
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet result = db.Get(new Car("BMW"));
				if (result.Size() < 1) 
				{
					return;
				}
				Car car = (Car)result[0];
				GenericReflector reflector = new GenericReflector(null,db.Ext().Reflector());
				ReflectClass carClass = reflector.ForObject(car);
				Console.WriteLine("Reflected class "+carClass);
				// public fields
				Console.WriteLine("FIELDS:");
				ReflectField[] fields = carClass.GetDeclaredFields();
				for (int i = 0; i < fields.Length; i++)
					Console.WriteLine(fields[i].GetName());
				
				// constructors
				Console.WriteLine("CONSTRUCTORS:");
				ReflectConstructor[] cons = carClass.GetDeclaredConstructors();
				for (int i = 0; i < cons.Length; i++)
					Console.WriteLine( cons[i]);
				
				// public methods
				Console.WriteLine("METHODS:");
				ReflectMethod method = carClass.GetMethod("ToString",null);
				Console.WriteLine(method.GetType());

			} 
			finally 
			{
				db.Close();
			}
		}
		// end GetCarInfo

		public  void GetReflectorInfo()
		{
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				Console.WriteLine("Reflector in use: " + db.Ext().Reflector());
				Console.WriteLine("Reflector delegate" +db.Ext().Reflector().GetDelegate());
				ReflectClass[] knownClasses = db.Ext().Reflector().KnownClasses();
				int count = knownClasses.Length;
				Console.WriteLine("Known classes: " + count);
				for (int i=0; i <knownClasses.Length; i++)
				{
					Console.WriteLine(knownClasses[i]);
				}
			} 
			finally 
			{
				db.Close();
			}
		}
		// end GetReflectorInfo
		
		public  void TestReflector()
		{
			LoggingReflector logger = new LoggingReflector();
			Db4o.Configure().ReflectWith(logger);
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				Car car = new Car("BMW");
				ReflectClass rc  = db.Ext().Reflector().ForObject(car);
				Console.WriteLine("Reflected class: " + rc);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end TestReflector

		public static void ListResult(ObjectSet result)
		{
			Console.WriteLine(result.Count);
			foreach (object item in result)
			{
				Console.WriteLine(item);
			}
		}
		// end ListResult
	}
}
